src/gettrail_print.c

```diff
@@ -0,0 +1,146 @@
+#define _XOPEN_SOURCE 700
+#define  _BSD_SOURCE
+
+#include <stdio.h>
+#include <traildb.h>
+#include <Judy.h>
+#include <stdbool.h>
+#include <string.h>
+
+#include "safeio.h"
+
+typedef uint64_t timestamp_t;
+
+void print_trails(char **traildb_paths, int num_paths, Pvoid_t cookies) {
+    printf("[\n");
+
+    for (int ti = 0; ti < num_paths; ti++) {
+        tdb *db = tdb_init();
+        tdb_error res = tdb_open(db, traildb_paths[ti]);
+        CHECK_TDB(res, "failed to open traildb %s", traildb_paths[ti]);
+
+        int num_fields = tdb_num_fields(db);
+
+        printf("{\n");
+
+        int num_trails = tdb_num_trails(db);
+        bool found_cookie = false;
+        for (int i = 0; i < num_trails; i++) {
+            const uint8_t *cookie = tdb_get_uuid(db, i);
+
+            uint8_t hexcookie[33] = {0};
+            tdb_uuid_hex(cookie, hexcookie);
+
+            Word_t *pv;
+            JSLG(pv, cookies, hexcookie);
+
+            if (pv) {
+                tdb_cursor *cursor = tdb_cursor_new(db);
+                tdb_error e = tdb_get_trail(cursor, i);
+                CHECK_TDB(e, "could not read trail %d", i);
+
+                const tdb_event *event;
+                if (!found_cookie) {
+                    printf("\"%s\": ", (const char *)hexcookie);
+                    found_cookie = true;
+                } else {
+                    printf(",\"%s\": ", (const char *)hexcookie);
+                }
+
+                printf("[\n");
+                int ei = 0;
+                while ((event = tdb_cursor_next(cursor))) {
+                    if (ei == 0) {
+                        printf("{\n");
+                    } else {
+                        printf(",{\n");
+                    }
+
+                    timestamp_t ts = (timestamp_t)event->timestamp;
+                    printf("\"timestamp\": %" PRIu64 ",\n", ts);
+
+                    for (int k = 1; k < num_fields; k++) {
+                        const char *field_name = tdb_get_field_name(db, k);
+                        uint64_t len;
+                        const char *field_value = tdb_get_item_value(db, event->items[k-1], &len);
+                        if(field_value != NULL) {
+                            printf("\"%s\": \"", field_name);
+
+                            // Print each character individually, skipping over quotes.
+                            const char *fvp = field_value;
+                            while (*fvp) {
```

Comment:
Why not iterate here using `len`? is the field value a null-terminated string?

Reply:
I guess I was assuming it was null-terminated, but it doesn't explicitly say so in the API docs: http://traildb.io/docs/api/#tdb_get_item_value

I also figured it'd look a bit weird to have a `for` loop where the loop control variable sometimes doesn't increment, but I could just use a while loop instead. Might be safer to just use `len`.

Reply:
Nevermind, a `for` loop looks just fine. I think I had a prior implementation of this that used two pointers to overwrite the string (but I couldn't actually do that because the original string was `const`), and this was a carryover from that.

Reply:
👍